sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1911,452 @@ def transform(self, X):
+            self._sparse_fit(X)
+        else:
+            self._dense_fit(X)
+
+        self._build_f()
+
+        return self
+
+    def _dense_transform(self, X, direction=True):
+        """Forward and inverse transform for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+
+        direction : bool, optional (default=True)
+            If True, apply forward transform. If False, apply
+            inverse transform.
+
+        Returns
+        -------
+        X : ndarray, shape (n_samples, n_features)
+            Projected data.
+        """
+        if direction:
+            func_transform = self._f_transform
+        else:
+            func_transform = self._f_inverse_transform
+        output_distribution = getattr(stats, self.output_distribution)
+
+        references = np.linspace(0, 1, self.n_quantiles, endpoint=True)
+
+        for feature_idx, f in enumerate(func_transform):
+            # older version of scipy do not handle tuple as fill_value
+            # clipping the value before transform solve the issue
+            if direction:
+                lower_bound_x = self.quantiles_[0, feature_idx]
+                upper_bound_x = self.quantiles_[-1, feature_idx]
+                lower_bound_y = references[0]
+                upper_bound_y = references[-1]
+            else:
+                lower_bound_x = references[0]
+                upper_bound_x = references[-1]
+                lower_bound_y = self.quantiles_[0, feature_idx]
+                upper_bound_y = self.quantiles_[-1, feature_idx]
+            if not direction:
+                #  for inverse transform, match a uniform PDF
+                for i in range(X.shape[0]):
+                    X[i, feature_idx] = output_distribution.cdf(
+                        X[i, feature_idx])
+            # Avoid computing for bounds due to numerical error of interp1d
+            lower_bounds_idx = (X[:, feature_idx] - BOUNDS_THRESHOLD <
+                                lower_bound_x)
+            upper_bounds_idx = (X[:, feature_idx] + BOUNDS_THRESHOLD >
+                                upper_bound_x)
+            bounds_idx = np.bitwise_or(lower_bounds_idx, upper_bounds_idx)
+            X[~bounds_idx, feature_idx] = f(X[~bounds_idx, feature_idx])
+            X[upper_bounds_idx, feature_idx] = upper_bound_y
+            X[lower_bounds_idx, feature_idx] = lower_bound_y
+            # for forward transform, match the output PDF
+            if direction:
+                for i in range(X.shape[0]):
+                    X[i, feature_idx] = output_distribution.ppf(
+                        X[i, feature_idx])
+                # find the value to clip the data to avoid mapping to
+                # infinity. Clip such that the inverse transform will be
+                # consistent
+                clip_min = output_distribution.ppf(BOUNDS_THRESHOLD / 10)
```

Comment:
is the `/10` for the sake of consistent inverse? I don't get why this magic number.